README
renameutils - Introduction
==========================

The file renaming utilities (renameutils for short) are a set of programs
designed to make renaming of files faster and less cumbersome.

The file renaming utilities consists of five programs - qmv, qcp, imv, icp
and deurlname.

The qmv ("quick move") program allows file names to be edited in a text
editor. The names of all files in a directory are written to a text file,
which is then edited by the user. The text file is read and parsed, and the
changes are applied to the files.

The qcp ("quick cp") program works like qmv, but copies files instead of
moving them.

The imv ("interactive move") program, is trivial but useful when you are
too lazy to type (or even complete) the name of the file to rename twice.
It allows a file name to be edited in the terminal using the GNU Readline
library. icp copies files.

The deurlname program removes URL encoded characters (such as %20
representing space) from file names. Some programs such as w3m tend to keep
those characters encoded in saved files.

The file renaming utilities are written in C by Oskar Liljeblad, and is
free software licensed under terms of the GNU General Public License.

Note that these programs come with no warranty whatsoever. They should be
considered beta quality although I have carefully tested them and verified
the code. This means that I can not be responsible for any data loss caused
by the file renaming utilities. If you are afraid that that is going to
happen, run qmv in dummy mode first (see below).

Copyright and License
=====================

renameutils is copyright (C) 2001, 2002, 2004, 2005, 2007, 2008
Oskar Liljeblad.

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc., 59
Temple Place, Suite 330, Boston, MA  02111-1307  USA

The source code of this project may contain files from other projects, and
files generated by other projects, including:

 * GNU Autoconf (http://www.gnu.org/software/autoconf/)
 * GNU Automake (http://sources.redhat.com/automake/)
 * GNU gettext (http://www.gnu.org/software/gettext/)
 * Gnulib (http://www.gnu.org/software/gnulib/)
 * Autoconf Macro Archive (http://ac-archive.sourceforge.net/)

Such files are licensed under the terms of the GNU General Public License
or a license compatible with the GNU GPL (as listed on
http://www.gnu.org/licenses/license-list.html). See each file for copyright
details.

The Swedish translations in the po directory may contain translations from
other projects, including:

 * GNU C Library (http://www.gnu.org/software/libc/libc.html)
 * GNU Core Utilities (http://www.gnu.org/software/coreutils/)

See the specific message file (PO file) for copyright of those messages.

Requirements
============

The following programs are required to build renameutils:

 * GCC 2.95 or later

   The GNU C Compiler is recommended, but other compilers might work.

 * GNU libintl

   This is needed for localisation. If you don't have it, you can use the
   --disable-nls option with the configure script.

 * GNU Coreutils

   A GNU version of the `ls' command is requred.

The following programs are required for proper operation of renameutils:

 * GNU Readline Library (libreadline) 2.1 or later

   Both qmv and imv use readline.

 * The unix `mv' command

   The imv program invokes `mv'.

 * GNU libintl

   This is needed unless you disable NLS support (see above).

Installation
============

Compile the renameutils by running `configure' and then `make'. This should
produce executables in the src directory, which can be used right away.

If you want to install the renameutils on your system, run `make install'.
This will copy the executables and the man pages into the appropriate
directories (usually /usr/local/bin and /usr/local/share/man/man1,
respecively).

For further details regarding `configure' and `make', see the INSTALL file.

Usage
=====

A list of options that each program accepts can be displayed by running it
with the --help option:

   qmv --help
   qcp --help
   imv --help
   icp --help
   deurlname --help

Please consult the manual page for each program for full usage information.
These documents can usually be viewed this way:

   man ./qcmd.1

or

   man -l qcmd.1

To rename a file using imv, use

   imv FILENAME

To rename files in the current directory using qmv, use

   qmv

or use the interactive mode of qmv:

   qmv -i

Enable the simulation mode of qmv by specifying the --dummy option. I use
imv and qmv in my user menu in Midnight Commander. The following lines are
in my ~/.mc/menu file:

     q	qmv
   	qmv .

     Q	qmv wide
   	qmv -owidth=120 .

     r	mv
   	imv %f

Homepage
========

The latest version of the file renaming utilities can be downloaded from

   http://www.nongnu.org/renameutils/

Feedback
========

The author of the file renaming utilities and this document, Oskar
Liljeblad, can be reached via email:

   oskar@osk.mine.nu (mailto:oskar@osk.mine.nu)

Please send bug reports, suggestions, ideas or comments in general to me.

Known bugs
==========

The imv program defaults to -i (--interactive) unlike mv which defaults to
-f (--force). This means that imv will ask before overwriting files.

The imv program trims the edited file name. This means that the destination
file name cannot have leading or trailing whitespace.

If you use alias mv=imv then you will only get the imv help and version
number with mv --help and mv --version.

Possible things to do
=====================

Please see the TODO document.

-


NEWS
2012-04-23: renameutils 0.12.0 released.
  Added `ls' variable and option to specify path of list program
  (suggested by Michael Williams).
  Fix bug with -d option, resulting in "skipping directory argument"
  error (reported by Stroller).
  Link with $(INTLLIBS) to fix Cygwin compilation (patch from Christian
  Franke).
  Fix minor Makefile nit - use $(DESTDIR) (see Gentoo bug #215954).
  Fix Readline history detection problem.
  Updated Gnulib.
  Add note that GNU Coreutils ls is requred (from Michael Williams).

2011-08-18: renameutils 0.11.0 released.
  Added a note about Solaris packages of renameutils on the web page
  (thanks to Steve Christensen).
  Updated French translation from Manuel Pegourie-Gonnard, thanks.
  Applied manpage typo patch from David Sastre, thanks.
  Fix a crash on exit with circular renames (reported by various people).
  Fix bug when a directory would be specified along with files on the
  command line (reported by Axel Beckert through Francois Marier, Debian
  bug #600411).

2008-09-21: renameutils 0.10.0 released.
  Change license to GPL v3.
  Added "import" command, suggested by Lou Afonso.
  Added "--command" option to specify command to use when renaming/copying
  files with qcmd, suggested by Nick Pope and Ulrich Dangel.
  Applied manual page lintian warning patch from Francois Marier.

2007-12-04: renameutils 0.9.0 released.
  Clean debian files in distclean-local, not clean-local.
  Do not remove debian/compat in distclean-local.
  Detect mv and cp options properly for icmd.
  Fix handling of trailing slashes in icmd (reported by Francois Marier).
  Use AC_SYS_LARGEFILE to fix renaming problems with large files (Debian
  bug report).
  Use program_name, not program_invocation_name for improved portability
  (bug reported by Andreas Eriksson).
  Various build and compatibility (Gnulib) fixes.
  Change icmd behavior when multiple arguments are specified - add option
  "--pass-though".
  Miscellaneous icmd improvements. Merge imv and icp manual page.
  Fix `plan' command crash and other crashes, reported by Mike Frysinger.
  Updated French and Swedish translations.

2005-08-23: renameutils 0.8.1 released.
  Now with complete French translation.

2005-08-23: renameutils 0.8.0 released.
  Removed debugging statements rendering imv and icp unusable (oops!)
  (reported by Francois Marier).
  Use mv(1) to perform the actual rename. Fixes Debian bug 316113
  (http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=316113). (Reported by
  Daniel Hulme through Francois Marier.)
  Improved some messages and error messages.
  Change order of inclusion of readline headers, fixing complation problem
  on some distributions (patch from Brian Hall).
  Added --command and --arg-options option to icmd (icp and imv).

2005-08-16: web site updated.
  Add some missing entries to this history.

2005-08-16: renameutils 0.7.0 released
  Added Swedish translation.
  Added French translation (from Francois Marier).
  The executables in the RPM packages now links with ncurses to prevent
  problems on Red Hat and Fedora systems.
  Handle POSIXLY_CORRECT in imv (fixes rename bug) (from Brian Hall).
  Change imv to pass command line through to mv if at least two file names
  are specified (or --target-directory option for mv) (suggested by
  Francois Marier).
  imv now checks if the source file exists prior to asking for destination
  name (suggested by Francois Marier).
  imv now detects and handles options requiring arguments for mv.
  imv now trims the edited file name. This means that the destination file
  name cannot have leading or trailing white space.
  Merge code base for imv and icp into icmd.
  Got rid of year ranges as recommended by the FSF.
  Improve the manual pages a little.
  Add Gnulib modules quote, dirname. Remove stat as it no longer exists in
  Gnulib.
  Fixed some typos in messages of the program.

2005-06-28: web site updated.
  Update description in README and web page.

2005-06-28: renameutils 0.6.0 released.
  Added new programs, icp and qcp (suggested by Mario Vanoni).
  Quote file names and strings before printing them (like GNU coreutils).
  Make use of more gnulib modules (quotearg, yesno, version-etc).
  Fixed a serious memory management-related bug in imv causing renaming to
  fail for no reason.
  Fixed a potential segmentation fault when printing a particular error
  message.
  Added copyright and license header to source files and manual pages.
  Manual page updates:
    Added manual page for deurlname.
    Corrected grammar.
    Removed section in imv about placing it in the Midnight Commander menu.
  Gettext is now external (no intl directory).
  Use Gnulib properly (modules.gnulib and autogen.sh removed).
  Added debian package.
  Added MANIFEST.sources file.
  Fix potential bug in cross-reference solution code.
  Added documentation for qcmd --help from Stephen Kolupaev, thanks.

2004-09-17: renameutils 0.5.1 released.
  Clarified README about requirements.
  Include Makefile.am.head in lib and m4 directories in distribution
  tarball.
  Minor clean ups to the autogen.sh, Makefile.am and configure.ac files.
  Now compiles with --disable-nls as well:
    Don't include (<gettext.h is enough).
    Include <locale.h> where setlocale is used.
  Check that locale was set correctly.

2004-08-26: renameutils 0.5.0 released.
  Added deurlname - a program to remove URL-escaped character sequences
  from filenames (such as %20 for space).
  Internationalized renameutils - added support for gettext.
  Removed debian support since renameutils has been in the official Debian
  archives for quite some time.
  Removed show-control-chars variable and option (automaticly assumed since
  we are using gettext).
  Fixed a bug in dual-column editformat that would cause a segfault when
  renaming files with trailing space in their names (reported by Chris
  Jantzen).
  Fixed a bug in qmv that made it impossible to rename files larger than

src/icmd.c
/* icmd.c - Interactive mv(1) and cp(1).
 *
 * Copyright (C) 2001, 2002, 2004, 2005, 2007, 2008 Oskar Liljeblad
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <config.h>
#include <unistd.h> 	    	/* gnulib (POSIX) */
#include <sys/stat.h>		/* POSIX */
#include <signal.h> 	    	/* gnulib (POSIX) */
#include <stdlib.h> 	    	/* C89 */
#include <stdio.h>  	    	/* C89 */
#include <errno.h>  	    	/* C89 */
#include <locale.h> 	    	/* gnulib (POSIX) */
#include <stdbool.h>	    	/* gnulib (POSIX) */
#include <readline/readline.h>	/* Readline */
#include <readline/history.h>	/* Readline */
#include <getopt.h>		/* gnulib (POSIX) */
#include <gettext.h>	    	/* gnulib (gettext) */
#define _(s) gettext(s)
#define N_(s) (s)
#include "configmake.h"		/* gnulib */
#include "progname.h"	    	/* gnulib */
#include "quotearg.h"	    	/* gnulib */
#include "quote.h"		/* gnulib */
#include "yesno.h"  	    	/* gnulib */
#include "version-etc.h"    	/* gnulib */
#include "xalloc.h"		/* gnulib */
#include "dirname.h"		/* gnulib */
#include "common/error.h"
#include "common/io-utils.h"
#include "common/string-utils.h"
#include "common/common.h"

#define MV_COMMAND "mv"
#define CP_COMMAND "cp"
/* This list should be up to date with mv and cp!
 * It was last updated on 2007-11-30 for
 * Debian coreutils 5.97-5.4 in unstable.
 *
 * Do not include -t, --target-directory in this list!
 *
 * If you update these lists, don't forget to update the manual page as well!
 */
#define MV_REQ_ARG_OPTIONS "S,reply,suffix"
#define CP_REQ_ARG_OPTIONS "S,Z,no-preserve,sparse,suffix,context"

const char version_etc_copyright[] = "Copyright (C) 2001, 2002, 2004, 2005, 2007, 2008 Oskar Liljeblad";

static char *first_text = NULL;

static int
insert_first_text(void)
{
    return rl_insert_text(first_text);
}

static void
int_signal_handler(int signal)
{
    puts("");
    warn(_("no changes made"));
    exit(EXIT_SUCCESS);
}

static void
display_help(void)
{
    printf(_("Usage: %s [OPTION] FILE...\n"), program_name);
    printf(_("Rename (imv) or copy a file (icp) by editing the destination name using\n"
             "GNU Readline. All options except those listed below are passed on to mv,\n"
             "cp, or the command specified by --command.\n\n"));
    printf(_("      --command=FILE         command to run instead of default mv/cp\n"));
    printf(_("      --arg-options=OPTIONS  list of options that require an argument\n"));
    printf(_("      --pass-through         run the command if two or more arguments are specified\n"));
    printf(_("      --help                 display this help and exit\n"));
    printf(_("      --version              output version information and exit\n"));
    printf(_("\nReport bugs to <%s>.\n"), PACKAGE_BUGREPORT);
}

/* Improve: return string position or word index in csv where value was found
 * XXX: move to strutil.c or something. Also in gmediaserver and microdc.
 */
bool
string_in_csv(const char *csv, char sep, const char *value)
{
    const char *p0;
    const char *p1;
    size_t len;
    
    len = strlen(value);
    for (p0 = csv; (p1 = strchr(p0, sep)) != NULL; p0 = p1+1) {
        if (p1-p0 == len && strncmp(p0, value, len) == 0)
            return true;                                 
    }
    if (strcmp(p0, value) == 0)
        return true;
    
    return false;
}

static void
trim(char *str)
{
    int h, t;

    for (t = strlen(str); t > 0 && isspace(str[t-1]); t--);
    str[t] = '\0';
    for (h = 0; isspace(str[h]); h++);
    if (h > 0)
        memcpy(str, str+h, t-h+1);
}

int
main(int argc, char **argv)
{
    struct sigaction action;
    bool no_more_opts = false;
    bool skip_opt_arg = false;
    bool posixly_correct;
    bool has_target_dir = false;
    bool pass_through = false;
    char *last_long_opt = NULL;
    char last_short_opt = '\0';
    char *arg_options = "";
    char *command = NULL;
    char *files[argc];
    /*char *opts[argc];*/
    int file_count = 0;
    /*int opt_count = 0;*/
    struct stat sb;
    char *srcfile;
    char *srcfile_stripped;
    char *program_base_name;
    int c;
    int first_cmd_opt;

    set_program_name(argv[0]);
    program_base_name = base_name(program_name);

    if (setlocale(LC_ALL, "") == NULL)
        warn(_("cannot set locale: %s\n"), errstr);
#ifdef ENABLE_NLS
    if (bindtextdomain(PACKAGE, LOCALEDIR) == NULL)
        warn(_("cannot bind message domain: %s\n"), errstr);
    if (textdomain(PACKAGE) == NULL)
        warn(_("cannot set message domain: %s\n"), errstr);
#endif

    if (strcmp(program_base_name, "imv") == 0) {
        arg_options = MV_REQ_ARG_OPTIONS;
        command = MV_COMMAND;
    } else if (strcmp(program_base_name, "icp") == 0)  {
        arg_options = CP_REQ_ARG_OPTIONS;
        command = CP_COMMAND;
    }

    for (c = 1; c < argc; c++) {
        char *arg = argv[c];

        if (strcmp(arg, "--help") == 0) {
            display_help();
            exit(EXIT_SUCCESS);
        } else if (strcmp(arg, "--version") == 0) {
            version_etc(stdout, base_name(program_name), PACKAGE, VERSION, "Oskar Liljeblad", NULL);
            exit(EXIT_SUCCESS);
        } else if (strcmp(arg, "--command") == 0) {
            if (++c >= argc)
                die(_("option `--%s' requires an argument"), arg+2);
            command = argv[c];
        } else if (strcmp(arg, "--arg-options") == 0) {
            if (++c >= argc)
                die(_("option `--%s' requires an argument"), arg+2);
            arg_options = argv[c];
        } else if (strcmp(arg, "--pass-through") == 0) {
            pass_through = true;
        } else if (strncmp(arg, "--command=", 10) == 0) {
            command = arg+10;
        } else if (strncmp(arg, "--arg-options=", 14) == 0) {
            arg_options = arg+14;
        } else {
            break;
        }
    }

    if (command == NULL && strcmp(program_base_name, "imv") != 0 && strcmp(program_base_name, "icp") != 0)
        die(_("must be started as either `imv' or `icp', or --command must be specified"));

    first_cmd_opt = c;

    memset(&action, 0, sizeof(sigaction));
    action.sa_handler = int_signal_handler;
    action.sa_flags = SA_RESTART;
    if (sigaction(SIGINT, &action, NULL) < 0)
	die(_("cannot register signal handler - %s\n"), errstr);
    action.sa_handler = SIG_IGN;
    if (sigaction(SIGQUIT, &action, NULL) < 0)
	die(_("cannot register signal handler - %s\n"), errstr);

    posixly_correct = getenv("POSIXLY_CORRECT") != NULL;

    for (c = first_cmd_opt; c < argc; c++) {
        char *arg = argv[c];
        int d;

        if (skip_opt_arg) {
            /*opts[opt_count++] = arg;*/
            skip_opt_arg = false;
        } else if (no_more_opts) {
            files[file_count++] = arg;
        } else if (arg[0] == '-' && arg[1] == '-' && arg[2] == '\0') {
            /*if (posixly_correct) {
                files[file_count++] = arg;
            } else {*/
                no_more_opts = true;
            /*}*/
        } else if (arg[0] == '-' && arg[1] == '-') {
            /*opts[opt_count++] = arg;*/
            if (strchr(arg, '=') == NULL) {
                if (string_in_csv(arg_options, ',', arg+2)) {
                    skip_opt_arg = true;
                    last_long_opt = arg+2;
                    last_short_opt = '\0';
                }
            }
            if (strncmp(arg, "--target-directory", 18) == 0)
                has_target_dir = true;
        } else if (arg[0] == '-') {
            /*opts[opt_count++] = arg;*/
            for (d = 1; arg[d] != '\0'; d++) {
                char option[2] = { arg[d], '\0' };
                if (string_in_csv(arg_options, ',', option)) {
                    skip_opt_arg = true;
                    last_long_opt = NULL;
                    last_short_opt = arg[d];
                    break;
                }
            }
        } else {
            files[file_count++] = arg;
            if (posixly_correct)
                no_more_opts = true;
        }
    }

    if (skip_opt_arg) {
        if (last_long_opt != NULL) 
            warn(_("option `--%s' requires an argument"), last_long_opt);
        else
            warn(_("option requires an argument -- %c"), last_short_opt);
	fprintf(stderr, _("Try `%s --help' for more information.\n"), program_name);
	exit(EXIT_FAILURE);
    }

    if (file_count == 0) {
	warn(_("missing file argument"));
	fprintf(stderr, _("Try `%s --help' for more information.\n"), program_name);
	exit(EXIT_FAILURE);
    }

    if (!pass_through && file_count > 1) {
        warn(_("too many arguments"));
        exit(EXIT_FAILURE);
    }

    if (file_count > 1 || has_target_dir) { /* Pass everything as is to command */
        char *args[argc-first_cmd_opt+3];

        memcpy(args+1, argv+first_cmd_opt, (argc-first_cmd_opt+2) * sizeof(char *));
        args[0] = command;
        if (execvp(command, args) < 0)
            die(_("cannot execute %s: %s"), command, errstr);
        exit(EXIT_FAILURE);
    }

    /* XXX: Here we should check mv version using 'mv --version', so that
     * we know the correct set of options for mv. Pseudocode:
     */
    /*if (opt_count > 0) {
        pipe
        fork
            in child
                close read end of pipe
                dup2 stdout, write end of pipe
                execlp(command, command, "--version", NULL);
        close write end of pipe
        getline from pipe

        if (strncmp(line, "mv (coreutils) ", 15) != 0)
            cannot check version. fail.
        if (strverscmp(line+15, COREUTILS_MAX_VERSION) >= 0)
            incorrect version. fail.

         process line
         while (getline != -1);
         if (ferror)
              report error
         waitpid
         cleanup
    }*/

    srcfile = files[0];
    if (lstat(srcfile, &sb) < 0)
        die(_("cannot stat %s: %s"), quote(srcfile), errstr);

    /* We usually don't want trailing slashes in the edited file name.
     * This is a matter of taste though.
     */
    srcfile_stripped = xstrdup(srcfile);
    strip_trailing_slashes(srcfile_stripped);

    rl_readline_name = base_name(program_name);
    first_text = xstrdup(srcfile_stripped);
    for (;;) {
        char *newname;

	add_history(srcfile_stripped);
	rl_startup_hook = insert_first_text;
	newname = readline("> ");

	if (newname != NULL) {
	    trim(newname);
	    strip_trailing_slashes(newname);

	    if (newname[0] != '\0' && strcmp(newname, srcfile_stripped) != 0) {
	        char *args[argc-first_cmd_opt+5];

                if (file_exists(newname)) {
                    printf(_("%s: overwrite %s? "), program_name, quote(newname));
                    fflush(stdout);
                    if (!yesno()) {
                        free(first_text);
                        first_text = newname;
                        continue;
                    }
                }

                free(srcfile_stripped);
                memcpy(args+1, argv+first_cmd_opt, (argc-first_cmd_opt+1) * sizeof(char *));
                if (posixly_correct || no_more_opts) {
                    args[argc-first_cmd_opt+1] = newname;
                    args[argc-first_cmd_opt+2] = NULL;
                } else {
                    args[argc-first_cmd_opt+1] = "--";
                    args[argc-first_cmd_opt+2] = newname;
                    args[argc-first_cmd_opt+3] = NULL;
                }
                args[0] = command;
                if (execvp(command, args) < 0)
                    die(_("cannot execute %s: %s"), command, errstr);
                exit(EXIT_FAILURE);
            }
	}

        free(first_text);
        if (newname == NULL)
            puts("");

        warn(_("no changes made"));
        free(srcfile_stripped);
        exit(EXIT_SUCCESS);
    }
}
